src/hayhooks/cli/utils.py

```diff
@@ -49,3 +63,186 @@ def make_request(
     except Exception as e:
         console.print(f"[red][bold]Unexpected error[/bold]\n{str(e)}[/red]")
         raise typer.Abort()
+
+
+def show_error_and_abort(message: str, highlight: str = "") -> None:
+    """Display error message in a panel and abort."""
+    if highlight:
+        message = message.replace(highlight, f"[red]{highlight}[/red]")
+    console.print(Panel.fit(message, border_style="red", title="Error"))
+    raise typer.Abort()
+
+
+def show_success_panel(message: str, title: str = "Success") -> None:
+    """Display success message in a panel."""
+    console.print(Panel.fit(message, border_style="green", title=title))
+
+
+def show_warning_panel(message: str, title: str = "Warning") -> None:
+    """Display warning message in a panel."""
+    console.print(Panel.fit(message, border_style="yellow", title=title))
+
+
+def with_progress_spinner(description: str, operation: Callable, *args, **kwargs):
+    """Execute an operation with a progress spinner.
+
+    Args:
+        description: Description to show in the spinner
+        operation: Function to call
+        *args, **kwargs: Arguments to pass to the operation
+
+    Returns:
+        The result of the operation
+    """
+    with Progress(
+        SpinnerColumn(),
+        TextColumn("[progress.description]{task.description}"),
+        console=console,
+    ) as progress:
+        progress.add_task(description=description, total=None)
+        return operation(*args, **kwargs)
+
+
+class ProgressFileReader:
+    """File-like object wrapper that updates progress bar when read."""
+
+    def __init__(self, file_obj, progress, task_id, file_size):
+        self.file_obj = file_obj
+        self.progress = progress
+        self.task_id = task_id
+        self.file_size = file_size
+        self.bytes_read = 0
+
+    def read(self, size=-1):
+        chunk = self.file_obj.read(size)
+        chunk_size = len(chunk)
+        self.bytes_read += chunk_size
+        self.progress.update(self.task_id, advance=chunk_size)
+        return chunk
+
+    def seek(self, offset, whence=0):
+        return self.file_obj.seek(offset, whence)
+
+    def tell(self):
+        return self.file_obj.tell()
+
+    def close(self):
+        return self.file_obj.close()
+
+
+def prepare_files_with_progress(files: Dict[str, Path], progress, task_id) -> Tuple[List[Tuple], List]:
+    """Prepare files for upload with progress tracking.
+
+    Args:
+        files: Dictionary mapping file keys to Path objects
+        progress: Rich Progress instance
+        task_id: ID of the progress task to update
+
+    Returns:
+        Tuple containing (files_list for requests, file_handles to close later)
+    """
+    # Initialize mimetypes
+    mimetypes.init()
+
+    files_list = []
+    file_handles = []
+
+    for file_key, file_path in files.items():
+        # Determine content type using mimetypes module
+        content_type, _ = mimetypes.guess_type(file_path)
```

Comment:
as a heads up if you look in our [FileTypeRouter](https://github.com/deepset-ai/haystack/blob/ed931b4c2b8b0d6ede81520aa9d89332062295b3/haystack/components/routers/file_type_router.py#L18) we had to supplement the available mime types especially surrounding file types like markdown and msg files. 

dC similarly had to supplement their mimetype detection because of this.